### ISSUE TITLE
chef-server-ctl backup on 12.2.0 HA

### ISSUE THREAD

**Issue by mikkoc** [chef/chef-server#539]
Running `chef-server-ctl backup` on the primary node of an HA Chef setup, version 12.2.0.
Using an S3 bucket for bookshelf.
Getting the following:
`
tar: /var/opt/opscode/drbd/data/bookshelf: Cannot stat: No such file or directory
`

Full output:

```
# chef-server-ctl backup
Starting Chef Server backup
WARNING:  Offline backup mode must stop your Chef server before continuing.  You can skip this message by passing a "--yes" argument. Do you wish to proceed? (y/N):
y
Bringing down the Chef Server
ok: down: nginx: 1s
ok: down: oc_bifrost: 0s
ok: down: oc_id: 0s
ok: down: opscode-chef-mover: 1s, normally up
ok: down: opscode-erchef: 0s
ok: down: opscode-expander: 0s
ok: down: opscode-expander-reindexer: 0s
ok: down: opscode-solr4: 1s
ok: down: rabbitmq: 0s
ok: down: redis_lb: 0s
Dumping Postgresql database to /tmp/chef_backup20150921-15374-ibt8v6/chef_backup-2015-09-21-14-54-22.sql
could not change directory to "/home/ec2-user"
ok: down: postgresql: 0s
Writing backup manifest
Creating backup tarball
tar: Removing leading `/' from member names
tar: /var/opt/opscode/drbd/data/bookshelf: Cannot stat: No such file or directory
tar: Exiting with failure status due to previous errors
Bringing up the Chef Server
ok: run: keepalived: (pid 1693) 435500s
ok: run: nginx: (pid 15544) 0s, normally down
ok: run: oc_bifrost: (pid 15559) 0s, normally down
ok: run: oc_id: (pid 15605) 0s, normally down
ok: run: opscode-chef-mover: (pid 15621) 0s
ok: run: opscode-erchef: (pid 15663) 0s, normally down
ok: run: opscode-expander: (pid 15684) 1s, normally down
ok: run: opscode-expander-reindexer: (pid 15706) 0s, normally down
ok: run: opscode-solr4: (pid 15734) 1s, normally down
ok: run: postgresql: (pid 15766) 0s, normally down
ok: run: rabbitmq: (pid 15803) 0s, normally down
ok: run: redis_lb: (pid 16021) 0s, normally down
Exporting tarball to /var/opt/chef-backup
Cleaning up /tmp/chef_backup20150921-15374-ibt8v6
Backup Complete!
```

Also, how should this tarball be restored on different machines (i.e. different IPs, etc)

**Comment by stevendanna** [comment]
Ideally we would check if bookshelf is enabled here:

https://github.com/chef/chef_backup/blob/master/lib/chef_backup/strategy/backup/tar.rb#L102-L106

To restore the tarball, you should be able to move it to a different machine and run `chef-server-ctl restore /path/to/backup` after installing chef-server.

**Comment by mikkoc** [comment]
@stevendanna external DB also needs to be taken into consideration.

I added the external RDS PostgreSQL database and now when I try to take a backup I get this:

```
# chef-server-ctl backup
Starting Chef Server backup
WARNING:  Offline backup mode must stop your Chef server before continuing.  You can skip this message by passing a "--yes" argument. Do you wish to proceed? (y/N):
y
Bringing down the Chef Server
ok: down: nginx: 1s
ok: down: oc_bifrost: 0s
ok: down: oc_id: 0s
ok: down: opscode-chef-mover: 0s, normally up
ok: down: opscode-erchef: 0s
ok: down: opscode-expander: 0s
ok: down: opscode-expander-reindexer: 0s
ok: down: opscode-solr4: 0s
ok: down: rabbitmq: 0s
ok: down: redis_lb: 0s
Dumping Postgresql database to /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql
chpst: fatal: unknown user/group: opscode-pgsql
ERROR: Something wen't terribly wrong, aborting backup
ERROR: Expected process to exit with [0], but received '111'
---- Begin output of /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql ----
STDOUT: 
STDERR: chpst: fatal: unknown user/group: opscode-pgsql
---- End output of /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql ----
Ran /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql returned 111
Cleaning up /tmp/chef_backup20150924-27122-1nwsv9e
Bringing up the Chef Server
ok: run: keepalived: (pid 1898) 63680s
ok: run: nginx: (pid 27259) 0s, normally down
ok: run: oc_bifrost: (pid 27274) 0s, normally down
ok: run: oc_id: (pid 27322) 0s, normally down
ok: run: opscode-chef-mover: (pid 27338) 1s
ok: run: opscode-erchef: (pid 27380) 0s, normally down
ok: run: opscode-expander: (pid 27402) 0s, normally down
ok: run: opscode-expander-reindexer: (pid 27424) 1s, normally down
ok: run: opscode-solr4: (pid 27450) 1s, normally down
ok: run: rabbitmq: (pid 27480) 1s, normally down
ok: run: redis_lb: (pid 27710) 1s, normally down
/opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.2.0.rc.0/lib/mixlib/shellout.rb:289:in `invalid!': Expected process to exit with [0], but received '111' (Mixlib::ShellOut::ShellCommandFailed)
---- Begin output of /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql ----
STDOUT: 
STDERR: chpst: fatal: unknown user/group: opscode-pgsql
---- End output of /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql ----
Ran /opt/opscode/embedded/bin/chpst -u opscode-pgsql /opt/opscode/embedded/bin/pg_dumpall > /tmp/chef_backup20150924-27122-1nwsv9e/chef_backup-2015-09-24-09-24-01.sql returned 111
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.2.0.rc.0/lib/mixlib/shellout.rb:276:in `error!'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/chef_backup-0.0.1.dev.4/lib/chef_backup/helpers.rb:53:in `shell_out!'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/chef_backup-0.0.1.dev.4/lib/chef_backup/strategy/backup/tar.rb:58:in `dump_db'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/chef_backup-0.0.1.dev.4/lib/chef_backup/strategy/backup/tar.rb:136:in `backup'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/chef_backup-0.0.1.dev.4/lib/chef_backup/runner.rb:28:in `backup'
	from (eval):30:in `block in load_file'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/omnibus-ctl-0.4.2/lib/omnibus-ctl.rb:190:in `call'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/omnibus-ctl-0.4.2/lib/omnibus-ctl.rb:190:in `block in add_command_under_category'
	from /opt/opscode/embedded/lib/ruby/gems/2.1.0/gems/omnibus-ctl-0.4.2/lib/omnibus-ctl.rb:661:in `run'
	from /opt/opscode/embedded/service/omnibus-ctl/chef-server-ctl:192:in `<main>'
```

Since I'm not getting much luck with the backup/restore feature, I tried the following:
- take EBS volume snapshot 
- take RDS snapshot 
- bring up a new cluster, with new EBS volume and RDS database created from the snapshots.
- touch `/var/opt/opscode/bootstrapped` on the primary node
- run `chef-server-ctl reconfigure` hoping it's smart enough to realise it's not a fresh install.
Unfortunately that's not the case, it still tries to create database:
```
[2015-09-24T09:14:08+00:00] FATAL: 

-----------------------------------------------------------------------
CSPG016: The Chef Server database named 'bifrost' already exists on the
         PostgreSQL server. Please remove it before proceeding.

         See https://docs.chef.io/error_messages.html#cspg016-database-exists
         for more information.
-----------------------------------------------------------------------

```

Any ideas?

**Comment by stevendanna** [comment]
> @stevendanna external DB also needs to be taken into consideration.

@marcparadise Just finished a research spike about what is required to make backup/restore more sensible in the face of remote data stores for all of the backend data services.  This is on our radar and will be fixed.

> run chef-server-ctl reconfigure hoping it's smart enough to realise it's not a fresh install. Unfortunately that's not the case, it still tries to create database:

This error is actually coming from our preflight validators, which appear to be using the previous_run data to determine if they need to run the various "already exists" checks.

https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_postgres_validator.rb#L186

The previous_run data gets populated if the file /etc/opscode/chef-server-running.json is on disk:

https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-cookbooks/private-chef/recipes/config.rb#L22-L38

If you restore this from the previous installation as well, I believe you'll be able to pass that check.

We should probably clean this up to either use a single method of determining "already bootstrapped" state or by making bootstrapping each component more idempotent.

**Comment by mikkoc** [comment]
I copied the chef-server-running.json and now the error is this (from a primary/bootstrap node):

```
[2015-09-24T10:21:13+00:00] FATAL: Attempt to create secrets on non-bootstrap node in an H/A topology, please copy /etc/opscode/* around instead.

================================================================================
Recipe Compile Error in /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/recipes/default.rb
================================================================================

SystemExit
----------
exit

Cookbook Trace:
---------------
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:182:in `exit'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:182:in `generate_hex_if_bootstrap'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:453:in `gen_secrets_default'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:357:in `gen_secrets'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:598:in `generate_config'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/recipes/config.rb:90:in `from_file'
  /opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/recipes/default.rb:38:in `from_file'

Relevant File Content:
----------------------
/opt/opscode/embedded/cookbooks/cache/cookbooks/private-chef/libraries/private_chef.rb:

175:        PrivateChef['backend_vips']
176:      end
177:  
178:      # guards against creating secrets on non-bootstrap node
179:      def generate_hex_if_bootstrap(chars, ha_guard)
180:        if ha_guard
181:          Chef::Log.fatal("Attempt to create secrets on non-bootstrap node in an H/A topology, please copy /etc/opscode/* around instead.")
182>>         exit 44
183:        end
184:        SecureRandom.hex(chars)
185:      end
186:  
187:      def generate_hash
188:        results = { "private_chef" => {} }
189:        default_keys = [
190:          "opscode_chef",
191:          "redis_lb",

[2015-09-24T10:21:13+00:00] ERROR: Running exception handlers
[2015-09-24T10:21:13+00:00] ERROR: Exception handlers complete
```

**Comment by stevendanna** [comment]
@mikkoc You'll likely want the entire contents of /etc/opscode from the backed up machine.  This way you get all the previously generate database secrets, etc.

**Comment by mikkoc** [comment]
Ignore my previous comment, it was running with `bootstrap = false`.
Changing to true does fix the issue. I'm able to create the new cluster successfully by using the old `chef-server-running.json` !

**Comment by mikkoc** [comment]
Actually, despite `chef-server-ctl ha-status` and `chef-server-ctl status` returning OK on all services, the Chef server (master) is in a broken state.
Here's some tail logs:

```
==> /var/log/opscode/opscode-erchef/current <==
2015-09-24_11:41:04.55860 [error] gen_server <0.1667.0> terminated with reason: socket_closed_unexpectedly in amqp_network_connection:handshake_recv/0 line 462

==> /var/log/opscode/opscode-erchef/crash.log <==
2015-09-24 11:41:04 =CRASH REPORT====
  crasher:
    initial call: amqp_network_connection:init/1
    pid: <0.1667.0>
    registered_name: []
    exception exit: {socket_closed_unexpectedly,[{gen_server,terminate,7,[{file,"gen_server.erl"},{line,804}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]}
    ancestors: [<0.1665.0>,gen_bunny_mon,gen_bunny_sup,<0.1535.0>]
    messages: []
    links: [<0.1665.0>,#Port<0.6333>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 675
  neighbours:

==> /var/log/opscode/opscode-erchef/erchef.log <==
2015-09-24 11:41:04.554 [error] CRASH REPORT Process <0.1667.0> with 0 neighbours exited with reason: socket_closed_unexpectedly in gen_server:terminate/7 line 804

==> /var/log/opscode/opscode-erchef/current <==
2015-09-24_11:41:04.56119 [error] CRASH REPORT Process <0.1667.0> with 0 neighbours exited with reason: socket_closed_unexpectedly in gen_server:terminate/7 line 804

==> /var/log/opscode/opscode-erchef/crash.log <==
2015-09-24 11:41:04 =SUPERVISOR REPORT====
     Supervisor: {<0.1665.0>,amqp_connection_sup}
     Context:    child_terminated
     Reason:     socket_closed_unexpectedly
     Offender:   [{pid,<0.1667.0>},{name,connection},{mfa,{amqp_network_connection,start_link,[{amqp_params,<<"actions">>,<<"0ccb0a1043a7d1b840034e869ed00578c79b97a3d48cc6278fa488eae1ef57006df999794484189607080abe48fa305665de">>,<<"/analytics">>,{127,0,0,1},5672,0,0,0,none,[]},<0.1666.0>,#Fun<amqp_connection_sup.0.94524864>]}},{restart_type,intrinsic},{shutdown,brutal_kill},{child_type,worker}]


==> /var/log/opscode/opscode-erchef/erchef.log <==
2015-09-24 11:41:04.555 [error] Supervisor {<0.1665.0>,amqp_connection_sup} had child at module undefined at <0.1667.0> exit with reason socket_closed_unexpectedly in context child_terminated

==> /var/log/opscode/opscode-erchef/current <==
2015-09-24_11:41:04.56352 [error] Supervisor {<0.1665.0>,amqp_connection_sup} had child at module undefined at <0.1667.0> exit with reason socket_closed_unexpectedly in context child_terminated

==> /var/log/opscode/opscode-erchef/crash.log <==
2015-09-24 11:41:04 =SUPERVISOR REPORT====
     Supervisor: {<0.1665.0>,amqp_connection_sup}
     Context:    shutdown
     Reason:     reached_max_restart_intensity
     Offender:   [{pid,<0.1667.0>},{name,connection},{mfa,{amqp_network_connection,start_link,[{amqp_params,<<"actions">>,<<"0ccb0a1043a7d1b840034e869ed00578c79b97a3d48cc6278fa488eae1ef57006df999794484189607080abe48fa305665de">>,<<"/analytics">>,{127,0,0,1},5672,0,0,0,none,[]},<0.1666.0>,#Fun<amqp_connection_sup.0.94524864>]}},{restart_type,intrinsic},{shutdown,brutal_kill},{child_type,worker}]


==> /var/log/opscode/opscode-erchef/erchef.log <==
2015-09-24 11:41:04.555 [error] Supervisor {<0.1665.0>,amqp_connection_sup} had child at module undefined at <0.1667.0> exit with reason reached_max_restart_intensity in context shutdown

==> /var/log/opscode/opscode-erchef/current <==
2015-09-24_11:41:04.56577 [error] Supervisor {<0.1665.0>,amqp_connection_sup} had child at module undefined at <0.1667.0> exit with reason reached_max_restart_intensity in context shutdown

==> /var/log/opscode/opscode-erchef/crash.log <==
2015-09-24 11:41:04 =ERROR REPORT====
["Could not connect, scheduling reconnect.",{error,{{error,{badmatch,{error,{auth_failure_likely,{socket_closed_unexpectedly,{gen_server,call,[<0.1667.0>,connect,infinity]}}}}},[{bunny_util,connect,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/bunny_util.erl"},{line,191}]},{gen_bunny_mon,do_connect,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,192}]},{gen_bunny_mon,handle_info,2,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,134}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]},{connection_info,{network,{127,0,0,1},5672,{<<"actions">>,<<"0ccb0a1043a7d1b840034e869ed00578c79b97a3d48cc6278fa488eae1ef57006df999794484189607080abe48fa305665de">>},<<"/analytics">>}}}}]

==> /var/log/opscode/opscode-erchef/erchef.log <==
2015-09-24 11:41:04.555 [error] "Could not connect, scheduling reconnect.", error: {{error,{badmatch,{error,{auth_failure_likely,{socket_closed_unexpectedly,{gen_server,call,[<0.1667.0>,connect,infinity]}}}}},[{bunny_util,connect,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/bunny_util.erl"},{line,191}]},{gen_bunny_mon,do_connect,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,192}]},{gen_bunny_mon,handle_info,2,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,134}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]},{connection_info,{network,{127,0,0,1},5672,{<<"actions">>,<<"0ccb0a1043a7d1b840034e869ed00578c79b97a3d48cc6278fa488eae1ef57006df999794484189607080abe48fa305665de">>},<<"/analytics">>}}}

==> /var/log/opscode/opscode-erchef/current <==
2015-09-24_11:41:04.56786 [error] "Could not connect, scheduling reconnect.", error: {{error,{badmatch,{error,{auth_failure_likely,{socket_closed_unexpectedly,{gen_server,call,[<0.1667.0>,connect,infinity]}}}}},[{bunny_util,connect,1,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/bunny_util.erl"},{line,191}]},{gen_bunny_mon,do_connect,3,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,192}]},{gen_bunny_mon,handle_info,2,[{file,"/var/cache/omnibus/src/oc_erchef/_build/default/lib/gen_bunny/src/gen_bunny_mon.erl"},{line,134}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,593}]},{gen_server,handle_msg,5,[{file,"gen_server.erl"},{line,659}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,237}]}]},{connection_info,{network,{127,0,0,1},5672,{<<"actions">>,<<"0ccb0a1043a7d1b840034e869ed00578c79b97a3d48cc6278fa488eae1ef57006df999794484189607080abe48fa305665de">>},<<"/analytics">>}}}
```

**Comment by stevendanna** [comment]
Provided that rabbitmq is properly started (chef-server-ctl status should list it as up and it should be listening on its TCP port), this appears to be a problem with the configured rabbitmq password.

**Comment by mikkoc** [comment]
This is the status.
I have not set any custom password for rabbitmq, simply copied the `chef-server-running.json` from the old installation.

```
# chef-server-ctl status
-------------------
 Internal Services 
-------------------
run: keepalived: (pid 4606) 214s; run: log: (pid 4621) 213s
run: nginx: (pid 4962) 173s, normally down; run: log: (pid 4593) 214s
run: oc_bifrost: (pid 4976) 171s, normally down; run: log: (pid 4421) 223s
run: oc_id: (pid 5021) 169s, normally down; run: log: (pid 4433) 222s
run: opscode-erchef: (pid 6009) 2s, normally down; run: log: (pid 4674) 210s
run: opscode-expander: (pid 5056) 164s, normally down; run: log: (pid 4675) 210s
run: opscode-expander-reindexer: (pid 5077) 162s, normally down; run: log: (pid 4571) 215s
run: opscode-solr4: (pid 5102) 160s, normally down; run: log: (pid 4570) 215s
run: rabbitmq: (pid 5131) 158s, normally down; run: log: (pid 4368) 224s
run: redis_lb: (pid 5464) 148s, normally down; run: log: (pid 4578) 215s
-------------------
 External Services 
-------------------
run: postgresql: connected OK to chef-test.xxxxxxxxx.eu-west-1.rds.amazonaws.com:5432
```

How to debug this further?

**Comment by mikkoc** [comment]
I managed to solve this by copying the whole `/etc/opscode/` folder as @stevendanna suggested.
In particular `private-chef-secrets.json` and `pivotal.pem` are needed.

**Comment by stevendanna** [comment]
Glad this worked out!  Apologies that the backup command doesn't yet work out of the box, we'll be working on that soon.

**Comment by dvinograd** [comment]
We have split Chef server up into sub-services that are run on multiple VMs - web app frontend, Solr, Postgres, and S3 instead of bookshelf.

In my case the issue was Chef bootstrap (on web app VMs) not creating rabbitmq accounts and rules during initial run. This caused rabbitmq to reject requests from opscode-erchef and other services and break the web app.

After much research it turned out to be because of `is_data_master` function returning `false` in `/opt/opscode/embedded/cookbooks/private-chef/recipes/rabbitmq.rb`. I traced down definition of `is_data_master` to `/opt/opscode/embedded/cookbooks/enterprise/libraries/helpers.rb`. To make it return `true` I've set the following attributes: `node['private_chef']['topology']="tier"` and `node['private_chef']['role']="backend"`. Subsequent runs of `chef-server-ctl reconfigure` on new and existing web app VMs started to evalued rabbitmq recipe and create rabbitmq users as needed.
